sisl/sparse.py

```diff
@@ -1323,24 +1358,35 @@ def tocsr(self, dim=0, **kwargs):
     def fromsp(cls, *sps, **kwargs):
         sps = list(map(lambda x: x.tocsr(), sps))
         shape = sps[0].shape
-        dtype = kwargs.get("dtype", np.result_type(*tuple(sp.dtype for sp in sps)))
-        # create a single sparse matrix to get the final size
-        # this will also error out if the shapes are not coherent
-        full_sp = reduce(lambda x, y: x + y, sps, 0.)
-        nnzpr = np.diff(full_sp.indptr).max() + 3
-        out = cls(shape + (len(sps), ), nnzpr=nnzpr, dtype=dtype)
-        del full_sp
-
-        # Now we need to add things to the sparsity pattern
-        for iD, sp in enumerate(sps):
-            ptr = sp.indptr
-            if sp.shape != shape:
-                raise ValueError(f"{cls.__name__}.fromsp found non compatible shapes")
-
-            # Loop stuff
-            for r in range(shape[0]):
-                idx = out._extend(r, sp.indices[ptr[r]:ptr[r+1]])
-                out._D[idx, iD] += sp.data[ptr[r]:ptr[r+1]]
+
+        if len(sps) == 1:
+            m = sps[0]
+            m.sort_indices()
+            out = cls(shape + (1,), nnzpr=1, nnz=1, dtype=m.dtype)
+            out.col = m.indices
+            out.ptr = m.indptr
+            out.ncol = np.diff(out.ptr)
+            out._nnz = len(out.col)
+            out._D = m.data.reshape(-1, 1)
+            if kwargs.get("copy", True):
```

Comment:
this is not good... :(

If a user passes a `csr_matrix` then `sps = list(map(lambda x: x.tocsr(), sps))` will be the same instance, and then you'll have two sparse matrices sharing `indices` and `indptr` in two different sparse matrices. Now if you change one of them, the other will too.

Copy should _always_ be true, regardless of what users may request, so better remove this...

So explicitly do `.copy()`.